rust/relay/src/sockets.rs

```diff
@@ -0,0 +1,259 @@
+use anyhow::{bail, Result};
+use std::{
+    collections::HashMap,
+    io,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+    task::{ready, Context, Poll},
+    time::Duration,
+};
+use stun_codec::rfc8656::attributes::AddressFamily;
+use tokio::sync::mpsc;
+
+const MAX_UDP_SIZE: usize = 65536;
+
+pub struct Sockets {
+    /// All currently active sockets.
+    ///
+    /// [`mio`] operates with a concept of [`mio::Token`]s so we need to store our sockets indexed by those tokens.
+    inner: HashMap<mio::Token, mio::net::UdpSocket>,
+
+    /// Which socket we should still be reading from.
+    ///
+    /// [`mio`] sends us a signal when a socket is ready for reading.
+    /// We must read from it until it returns [`io::ErrorKind::WouldBlock`].
+    current_ready_socket: Option<mio::Token>,
+
+    cmd_tx: mpsc::UnboundedSender<Command>,
+    event_rx: mpsc::Receiver<Event>,
+
+    buffer: [u8; MAX_UDP_SIZE],
+}
+
+impl Default for Sockets {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Sockets {
+    pub fn new() -> Self {
+        let (cmd_tx, cmd_rx) = mpsc::unbounded_channel();
```

Comment:
I still don't like the unbounded channels. If we know it can't grow without bound, then we could bound it. I'm not sure why they even offer it as an API. Just set the bound to 4 billion? Is bounds checking expensive?

Reply:
I can think of a bounded solution.

Reply:
I am now failing the entire relay if this ever fails.